### ISSUE TITLE
How to add success to every responce?

### ISSUE THREAD

**Issue by Sogl** [dingo/api#1224]
Hello!

I want to add `success` responce to every API call. If we have an error, `success: false` and if all is OK, then `success: true`.

Final responce must looke like this:
```
{
"success":true,
"data": [{}, {}]
}
```

How to do that?

**Comment by ahmadbadpey** [comment]
@Sogl,I have same problem , Have you reached a result?

**Comment by johannesschobel** [comment]
I guess (!!!), that you need to write your own transformer.. Take a look at this wiki entry here!
[Custom Transformers](https://github.com/dingo/api/wiki/Transformers#custom-transformation-layer)

It may help you?!

**Comment by Sogl** [comment]
@johannesschobel All transfomer manipulation add lines into `"data"` JSON section, but I need `"success"` near this section.

**Comment by johannesschobel** [comment]
yeah, but would that be a problem?
Your JSON response would look like this:

```json
{
   "data" : {
      "success" : true,
      "foo" : "bar",
   }
}
```

another way of doing this, would be to use the ```setMeta()``` function of the response. Using this method, you can apply a ```meta``` block. In this block, for example, pagination information (what page are you on, how many items are displayed on one page, ...) is stored..

You can simply append your own meta information either by using ```setMeta()``` (overwrite the meta block) or by using ```addMeta()``` (append to meta block)...

**Comment by ahmadbadpey** [comment]
>  but would that be a problem?

@johannesschobel , To simplify More. in most of major API that I checked there is comes a simple key like ok or success with a boolean value that shown result status In short. in the other hand, front-end developer has a simpler task too

**Comment by Sogl** [comment]
@johannesschobel So... I use `setMeta` now where pagination also present and in my frontend framework I set check `meta.success`. 

But things gets complicated. Why I can't simple change server responce in my own way, without hacking?
I need to put the same code in every responce... can I do that in one place when success and no errors?

**Comment by johannesschobel** [comment]
@ahmadbadpey and @Sogl 
that is - at least in my opinion - not good practice, because you should use HTTP status codes for this case.. e.g. ```200``` means ```ok``` (or in your case ```success```), if there is an error in the request (e.g., the requestor asks for an entity that is not available) you should throw a ```404``` (not found) and so on!

You are trying to "rebuild" functionality, which is already available through common HTTP standards. Why not check the HTTP status of the response?!

**Comment by Sogl** [comment]
It's additional data, not for replace status codes. 
ExtJS frontend framework need that data.

**Comment by johannesschobel** [comment]
then put it in the ```meta``` block, as it does **not** belong to your original data you want to transfer!

**Comment by Sogl** [comment]
Question was "how to put new data near the `data` block?", not into `meta` block.

**Comment by hskrasek** [comment]
@Sogl how are you letting the response(s) be generated? Just returning models and letting the framework / dingo handle it?

**Comment by Sogl** [comment]
@hskrasek I don't understand what you talking about.. can you rephrase, please?
I explained what I want in the 1st message.

**Comment by tantam** [comment]
You can follow this article https://github.com/dingo/api/wiki/Responses#morphing-and-morphed-events
Listen to the event and check their status code (or whatever data that you defined) and add what you want to response, example with "AddPaginationLinksToResponse" is clear and you can follow it

**Comment by tosslife** [comment]
i have the same question , Have you reached a result?

**Comment by anil1712** [comment]
I do have some doubt, is anyone knows how we can achieve this?

**Comment by cfeeling** [comment]
I implemented it as @tantam  already said. I prefer to put my implementation here for those person who was searching the solution.

Step 1. Create an event listener in your ```app/Listeners```, it was named AddWrapToResponse in my example:

```php
use Dingo\Api\Event\ResponseWasMorphed;

class AddWrapToResponse
{
    public function handle(ResponseWasMorphed $event)
    {
        if ($event->response->isSuccessful())
        {
            $data = $event->response->content();
            $event->response->setContent(json_encode([
                'code' => 0,
                'success' => TRUE,
                'data' => $data,
            ]));
        }
    }
}
```

Step 2. Register the created event listener into your ```app/Providers/EventServiceProvider`` as below:

```php
use Illuminate\Foundation\Support\Providers\EventServiceProvider as ServiceProvider;
use App\Listeners\AddWrapToResponse;

class EventServiceProvider extends ServiceProvider
{
    /**
     * The event listener mappings for the application.
     *
     * @var array
     */
    protected $listen = [
        'Dingo\Api\Event\ResponseWasMorphed' => [
            AddWrapToResponse::class,
        ]
    ];
}
```

After these two steps, all response from your dingo api would be enveloped with your custom data.

**Comment by beaplat-61f** [comment]
@Sogl You can use
```
return $this->response->array(['success' => true, 'data' => $data]);
```
